kinoml/modeling/homology.py

```diff
@@ -0,0 +1,272 @@
+            temp_file.seek(0)  # rewind the file for reading
+            # TODO add option to save original alignment
+
+            ali_lines = []
+            for line in temp_file.readlines():
+                line_str = line.decode("utf-8")
+                ali_lines.append(line_str.strip())
+
+            # split the list for easy reading
+            index = ali_lines.index(">P1;target_seq")
+            ali_1 = ali_lines[:index][1:-1]  # template
+            ali_2 = ali_lines[index:]  # target
+
+            ali_1_new, ali_2_new = [], []
+            ali_1_final, ali_2_final = [], []
+
+            # remove long blank regions in template seq "-"
+            for i, (a, b) in enumerate(zip(ali_1, ali_2)):
+
+                if any(line_element.isalpha() for line_element in a):
+                    ali_1_new.append(ali_1[i])
+                    ali_2_new.append(ali_2[i])  # remove corresponding seq in target
+
+            #  TODO change sequence numbers in alignment file,
+            # e.g. aln[0].range = ['3:', ':100']
+
+            for i, (line_ali_1, line_ali_2) in enumerate(zip(ali_1_new, ali_2_new)):
+                if i > 1:  # ignore the title lines, only focus on sequence
+
+                    index_dict = dict.fromkeys(np.arange(len(line_ali_1)).tolist())
+
+                    for i in range(len(line_ali_1) - window + 1):
+                        gap_count = 0
+                        chunk1 = line_ali_1[i : i + window]
+                        chunk2 = line_ali_1[i + window : i + (2 * window)]
+
+                        # check if there are '-'s in the chunk
+                        if not any(c.isalpha() for c in chunk1):
+                            for c1, c2 in zip(chunk1, chunk2):
+
+                                # check each character in chunk1 & chunk2
+                                if re.match("-", c1) and re.match("-", c2) is not None:
+                                    gap_count += 1
+
+                            # if we have a large region of '-'s, mark for deletion
+                            if gap_count == window:
+                                for key in np.arange(i, i + (2 * window)):
+                                    index_dict[key] = "delete"
+
+                    new_template_string = []
+                    for i, s in enumerate(line_ali_1):
+                        if index_dict[i] is None:
+                            new_template_string.append(line_ali_1[i])
+
+                    ali_1_final.append("".join(new_template_string))
+
+                    new_target_string = []
+                    for i, s in enumerate(line_ali_1):
+                        if index_dict[i] is None:
+                            new_target_string.append(line_ali_2[i])
+
+                    ali_2_final.append("".join(new_target_string))
+
+                else:
+                    continue
+
+        with open(self.alipath, "w") as ali_file:  # saving the file to cache
```

Comment:
File IO should happen in temporary directories, especially if they are one-time uses. This is super useful also if you need to call external binaries!

* MDTraj context manager: https://github.com/mdtraj/mdtraj/blob/master/mdtraj/utils/contextmanagers.py#L38-L46
* Example in openmoltools: https://github.com/choderalab/openmoltools/blob/master/openmoltools/amber.py#L181